### PR TITLE
Prometheus: Fix running of exemplar queries for non-histogram metrics

### LINKED ISSUE #42566: Prometheus: could not query exemplars for non histogram metrics
**What happened**:
When querying metrics like "increase(provider_failure_count_total{}[1m])" , I can't query with exemplars.

**What you expected to happen**:
I can query with exemplars for any metric.

**How to reproduce it (as minimally and precisely as possible)**:
I found the limit code in "public/app/plugins/datasource/prometheus/datasource.ts" relate to PR #39531 ，but I don't understand the reason "exemplars are available only for histogram metrics", as I known prometheus doesn't has this limit and I'm storing and querying exemplars with this metric "provider_failure_count_total" in prometheus now.

**Environment**:
- Grafana version: 8.3.0-beta2
- Data source type & version: Prometheus 2.31.1

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR fixes running of exemplar queries for non-histogram metrics. We should only check if we've already created an exemplar query for the same metric. If so, in that case don't run duplicated exemplars queries. This is a logic that should be in shouldRunExemplarQuery : https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/datasource.ts#L252-L264

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/42566

**How to review:**
1. Create a query that doesn't include histogram metric (e.g. `up`) with exemplars toggled on - exemplar in request should be `true`.
image

2. Create a query that includes 2 queries with the same histogram metric (e.g. `prometheus_http_response_size_bytes_bucket{}`) with exemplars toggled on - exemplar in the first request should be `true`, in the second false.

image

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 1]
```diff
@@ -300,29 +300,26 @@ export class PrometheusDatasource extends DataSourceWithBackend<PromQuery, PromO
...
-
-      if (histogramMetrics.length > 0) {
-        return !!histogramMetrics.find((metric) => target.expr.includes(metric));
-      } else {
-        return target.expr.includes('_bucket');
+      // We check all already processed targets and only create exemplar target for not used metric names
+      const metricName = this.languageProvider.histogramMetrics.find((m) => target.expr.includes(m));
+      const targets = [...request.targets];
+      // Remove targets that weren't processed yet (in targets array they are after current target)
+      targets.splice(request.targets.findIndex((t) => t.refId === target.refId));
```
Comment (gabor) [inline public/app/plugins/datasource/prometheus/datasource.ts:309]: can that `findIndex` fail? i mean, is it possible that it does not find that item and returns `-1`?

Comment (gabor) [reply public/app/plugins/datasource/prometheus/datasource.ts:309]: maybe we could change the
```js
const targets = [...request.targets]
targets.splice(something)
```

to

```js
const targets = request.targets.slice(0, something);
```
?
what do you think?

**Review by gabor — COMMENTED** [review]
just did a quick check of the code, will take a deeper look now.

**Review by gabor — APPROVED** [review]
LGTM (there are two comments i left in the previous review, but they are only small things)

File: public/app/plugins/datasource/prometheus/datasource.ts (hunk 1 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/datasource.ts]: yeah, good suggestion. fixed in 04a2b49

Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/datasource.ts]: > can that findIndex fail

No, tI don't this there can be a situation when this fails

**Comment by dbluxo** [comment]
Hi @ivanahuckova can it be that through this PR it comes to [this](https://github.com/grafana/grafana/issues/43323) issue?

We would now have to turn off the exemplar button in all dashboards for all queries, so that you don't get any more error messages, is that the way it's supposed to be? I have the same problem with version 8.3.4.

**Comment by ivanahuckova** [comment]
Yes, that is currently correct behaviour. We default to `exemplars: true` for discoverability of the feature, but it then means that if you don't have exemplars implemented for those queries, they result in error. There is this discussion/feature request for disabling exemplars globally https://github.com/grafana/grafana/discussions/41903. Would that solve your usecase? Or would you prefer to default to `exemplars: false` and toggle it only for specific queries?

**Comment by dbluxo** [comment]
Ok, thanks for the explanation.

I'd prefer to set it per default to `false` and only toggle it for specific queries.

**Comment by ivanahuckova** [comment]
Could I please ask you to create a new discussion https://github.com/grafana/grafana/discussions/new with this feature request? Thank you!

**Comment by dbluxo** [comment]
Done https://github.com/grafana/grafana/discussions/44264 :white_check_mark:

**Comment by ivanahuckova** [comment]
Awesome thanks!

**Pull request merged**